### ISSUE TITLE
Bug: Cost Reduction/Permission is using wrong player when card is not cast by its owner

### ISSUE THREAD

**Issue by Susucre** [magefree/mage#11166]
#11165 uncovered a bunch of bugged interaction between cards that allow to cast cards owned by another player (like [[Gonti, Lord of Luxury]]) and cost reduction/permission that is checking/counting something variable depending on who the caster is.

Some of the impacted cards:
- cost modification (variable or not) depending on who "you" is: [[The Great Henge]] [[Ghalta, Primal Hunger]] and other. But [[Academy Journeymage]] is also not checking if there is a wizard for the caster, but the card's owner instead.
- restriction that care about what you control/own. For instance the legendary sorcery/instant can not currently be cast with Gonti if the card's owner does not control a legend.

A few of the weird ones that were confirmed bugged [[Animist's Might]] [[Blizzard]]
A few I did not test, but no reason to suspect they are not bugged too: [[Bedlam Reveler]] [[Dargo, the Shipwrecker]] [[Korvold, Gleeful Glutton]]

**Comment by Susucre** [comment]
[[Serra Avenger]]

**Comment by Susucre** [comment]
[[Jaya's Immolating Inferno]]

**Comment by Susucre** [comment]
[[The Skullspore Nexus]] and [[The Great Henge]] will have the correct behavior next update.
It is actually a regression on the great henge with V6-beta2 to be wrong. Bug was reintroduced in an attempt to simplify the code, and we lacked an unit test there.

**Comment by JayDi85** [comment]
Nope, all dynamic value calculations go from source's controller side -- it's correct.

**Comment by Susucre** [comment]
@JayDi85 well something is wrong somewhere.

Try TheSkullsporeNexusTest::test_costreduction_opp_card with a stop in TheSkullsporeNexusReductionEffect::apply ; the second time we go through it, here are the parameters:

source (the controller Id is the card's owner):
![Image](https://github.com/user-attachments/assets/828ee2be-6c39-40e2-b418-b5a0626aa73e)

abilityToModify (the controller Id is the one that stole the card in the test, and is attempting a cast):
![Image](https://github.com/user-attachments/assets/b03de4b1-144e-47cb-bbce-99dbc1edb1d7)

If I wanted to replace the reduction computation by a dynamic value with an inner controlled filter, the test would be broken with source.

It might be an issue with how source is set up above.

**Comment by Susucre** [comment]
Okay, thinking more about it, source is right.
But there is an issue with cards with cost modification/condition on them, when they are not cast by their owner, as the modifier/condition has not the right controllerId.
When they are collected, they are not properly set in ContinuousEffects->costModification:

![Image](https://github.com/user-attachments/assets/aa9365a4-535b-4602-9811-e1e5545b41f9)

I am not sure if the engine should copy + set controller of the abilities on the card there, or if the issue is upstream in CostModificationEffect maintenance when someone is attempting to cast a card they don't own.

**Comment by JayDi85** [comment]
Looks like something wrong with linked abilities -- cause source's controller must be changed too, but it doesn't -- "source" must be also controlled by player A in that example:
![Image](https://github.com/user-attachments/assets/8187cbca-1192-4055-8724-978374ba9eae)

Controller changes on cast -- it also try to change `subabilities` (linked abilities) too:
![Image](https://github.com/user-attachments/assets/b694d2f4-4fe1-4e29-8a13-bf060691d115)
![Image](https://github.com/user-attachments/assets/d0e7b7af-e027-4160-a30f-fc491887b552)

It must temporary change and restore controller of all source object's abilities, not casting only -- like that:
![Image](https://github.com/user-attachments/assets/eb8def64-88cf-4f29-b197-720a2bb822d6)

**Comment by JayDi85** [comment]
I'm talking about code like that, but with try/finally and runtime checks for same controllerId in all abilities (to find out bad use cases with diff controllersId in diff abilities of the same card -- is it possible or not):

![Image](https://github.com/user-attachments/assets/08bb4fe6-341e-4c34-ba3d-99b120180b24)

**Comment by JayDi85** [comment]
That code example works fine in your use case and other tests but need deeper research for same places and potential side effects (can it lost data consistence or it's safe for any code conditions/restore/rollback).

**Comment by JayDi85** [comment]
Or maybe you don't need to restore it at all -- e.g. "if player cast card then all card's abilities become under their control too" -- so it can be added in card's cast method like that (also works fine in your use case and current tests and also require some research for side effects/tests):

![Image](https://github.com/user-attachments/assets/ca1462ac-f8a1-48a1-bc42-f34d931e5fa4)
